From 6c6656d0f78f1b68956e21ffa91dcfc2d94c808b Mon Sep 17 00:00:00 2001
From: Chao Guo <chao.guo@nxp.com>
Date: Fri, 14 Jun 2024 10:53:43 +0900
Subject: [PATCH 75/91] LF-12583 glupload: Add formats supported by
 #GstGLMemory to raw caps when generating sink pad caps

When glupload generates sink caps based on src caps, src caps may only contain RGBA format.
In this case, the raw caps on the sink pad generated by glupload will only contain the RGBA
format, which will cause caps negotiation fail, because the filter caps used for negotiation
by the upstream element may only contain other formats, such as xBGR, etc.

Add the formats supported by #GstGLMemory to raw caps to ensure that caps negotiation succeeds.

Upstream-Status: Pending
part-of: <https://gitlab.freedesktop.org/gstreamer/gstreamer/-/merge_requests/7061>

Signed-off-by: Chao Guo <chao.guo@nxp.com>
Signed-off-by: Xiaolei Wang <xiaolei.wang@windriver.com>
---
 gst-libs/gst/gl/gstglupload.c | 22 ++++++++++++++++++----
 1 file changed, 18 insertions(+), 4 deletions(-)

diff --git a/gst-libs/gst/gl/gstglupload.c b/gst-libs/gst/gl/gstglupload.c
index afdfd77..a55c1e0 100644
--- a/gst-libs/gst/gl/gstglupload.c
+++ b/gst-libs/gst/gl/gstglupload.c
@@ -1935,7 +1935,7 @@ _dma_buf_upload_free (gpointer impl)
 
 static const UploadMethod _dma_buf_upload = {
   "Dmabuf",
-  0,
+  METHOD_FLAG_CAN_ACCEPT_RAW,
   &_dma_buf_upload_caps,
   &_dma_buf_upload_new,
   &_dma_buf_upload_transform_caps,
@@ -2080,7 +2080,7 @@ _direct_dma_buf_upload_transform_caps (gpointer impl, GstGLContext * context,
 
 static const UploadMethod _direct_dma_buf_upload = {
   "DirectDmabuf",
-  0,
+  METHOD_FLAG_CAN_ACCEPT_RAW,
   &_dma_buf_upload_caps,
   &_direct_dma_buf_upload_new,
   &_direct_dma_buf_upload_transform_caps,
@@ -2102,7 +2102,7 @@ _direct_dma_buf_external_upload_new (GstGLUpload * upload)
 
 static const UploadMethod _direct_dma_buf_external_upload = {
   "DirectDmabufExternal",
-  0,
+  METHOD_FLAG_CAN_ACCEPT_RAW,
   &_dma_buf_upload_caps,
   &_direct_dma_buf_external_upload_new,
   &_direct_dma_buf_upload_transform_caps,
@@ -3822,8 +3822,22 @@ gst_gl_upload_transform_caps (GstGLUpload * upload, GstGLContext * context,
         GstCapsFeatures *passthrough =
             gst_caps_features_from_string
             (GST_CAPS_FEATURE_META_GST_VIDEO_OVERLAY_COMPOSITION);
-        GstCaps *raw_tmp = _set_caps_features_with_passthrough (tmp,
+        GstCaps *raw_tmp = _set_caps_features_with_passthrough (caps,
             GST_CAPS_FEATURE_MEMORY_SYSTEM_MEMORY, passthrough);
+
+        /* Add formats supported by #GstGLMemory to raw caps */
+        for (i = 0; i < gst_caps_get_size (raw_tmp); i++) {
+          GstStructure *s = gst_caps_get_structure (raw_tmp, i);
+          GValue formats = G_VALUE_INIT;
+
+          g_value_init (&formats, GST_TYPE_LIST);
+          gst_value_deserialize (&formats, GST_GL_MEMORY_VIDEO_FORMATS_STR);
+          gst_structure_take_value (s, "format", &formats);
+
+          gst_structure_remove_fields (s, "texture-target", NULL);
+          g_value_unset (&formats);
+        }
+
         gst_caps_append (tmp, raw_tmp);
         gst_caps_features_free (passthrough);
       }
-- 
2.43.0

